[FLINK-17467] Align channels on savepoint in UC mode

rkhachatryan opened the pull request:
## What is the purpose of the change

Perform alignment for savepoints even if UnalignedCheckpoints are enabled.

## Changelog

1. Introduce `AlternatingCheckpointBarrierHandler` which delegates to either Aligned or Unaligned barrier handlers
1. Don't notify barrier listener about savepoints in Input Channels
1. Don't include channel state into savepoint snapshots
1. Block upstream on savepoints in UC mode too

## Verifying this change

Added unit tests:
 - `AlternatingCheckpointBarrierHandlerTest`
 - `CheckpointOptionsTest`
 - `testNoNotifyOnSavepoint` for input channels
 - `SubtaskCheckpointCoordinatorTest.testSkipChannelStateForSavepoints`

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: yes
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? not applicable

AHeise commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/RemoteInputChannel.java:
```diff
@@ -575,12 +574,10 @@ public void onBuffer(Buffer buffer, int sequenceNumber, int backlog) throws IOEx
 				enqueued = true;
 
-				if (listener != null && buffer.isBuffer() && receivedCheckpointId < lastRequestedCheckpointId) {
-					notifyReceivedBuffer = buffer.retainBuffer();
```
Even if you go optional, I wouldn't go ternary here. Ternary is meant for some simple condition and one-liners.

rkhachatryan replied:
IMHO ternary is meant for branches that have only one "effect", usually assignment. And this is exactly the case here. 
`If-else` OTH is meant for cases when "effect" is something more complex.

If the condition is complex, it should be extracted to a method, regardless of whether it's `if-else` or ternary operator.

Here, I think the condition is fine (and I didn't change it).

pnowojski replied:
As before, it makes the diff/commit/PR more difficult to understand.

rkhachatryan replied:
Reverted to minimize changes in this PR.

AHeise commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/InputProcessorUtil.java:
```diff
@@ -108,18 +108,24 @@ private static CheckpointBarrierHandler createCheckpointBarrierHandler(
-					return new CheckpointBarrierUnaligner(
-						numberOfInputChannelsPerGate.toArray(),
-						channelStateWriter,
+				return config.isUnalignedCheckpointsEnabled() ?
```
Again ternary is not making anything easier here.

rkhachatryan replied:
Replied [above](https://github.com/apache/flink/pull/11948#discussion_r418034387).

rkhachatryan replied:
Sorry, I didn't get you point here, can you  please explain
> this change is +/- 0 and doesn't add/remove value

?

(the change is about using `AlternatingCheckpointBarrierHandler`, not ternary instead of if)

pnowojski replied:
Yes, the commit makes sense and it has value (using `AlternatingCheckpointBarrierHandler `). 

I meant that changing from if/else to ternary doesn't add or remove much value, it could be argued both ways which option is better, so changing it one way or another in an existing code is a bit pointless, even in separate dedicated commits. Doing it squashed with other changes (especially functional changes) is harmful.

rkhachatryan replied:
Reverted to minimize changes in this PR.

AHeise commented on flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/io/AlternatingCheckpointBarrierHandlerTest.java:
```diff
@@ -0,0 +1,212 @@
+		for (InputChannel channel : gate.getInputChannels().values()) {
+			assertFalse(barrierHandler.isBlocked(channel.getChannelIndex()));
+			assertEquals(
+					String.format("channel %d should be resumed", channel.getChannelIndex()),
```
nit: indent

AHeise commented on flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/io/AlternatingCheckpointBarrierHandlerTest.java:
```diff
@@ -0,0 +1,212 @@
+/**
+ * {@link AlternatingCheckpointBarrierHandler} test.
+ */
+public class AlternatingCheckpointBarrierHandlerTest {
```
Good tests! But I think you also need to test `BufferReceivedListener` as the current version of AlternatingCheckpointBarrierHandler is probably not allowing any UC on input side.

rkhachatryan replied:
Thanks! I'll test that `InputChannel`s don't notify `BufferReceivedListener`.

AHeise replied:
I was more concerned that the alternation is currently not working when a barrier is received through notify. In particular, an UC is only triggered when the first barrier is processed vs. received defeating the purpose of UC.

rkhachatryan replied:
`AlternatingCheckpointBarrierHandler` will receive the barrier as usual (via "`pollNext`") and switch its aligner. I don't see any reason to notify it immediately (or through the mailbox) as soon as the channel receives a barrier.

Do you see any problems with this approach?

pnowojski replied:
There is an inconsistency issue from perspective of InputChannels threads. But they are accessing `AlternatingCheckpointBarrierHandler` only via the `notifyBarrierReceived` and `notifyBufferReceived` and those methods are only called/active AFTER unaligned checkpoint has started in the task thread, so barrier had to be already processed by the `AlternatingCheckpointBarrierHandler`. Hence `activeHandler` had to already switch `unalignedHandler`.

This is very fragile contract, but :

- solution would require to synchronize code even more 
- I think there is no bug at the moment
- once we fix the threading model and get rid of the listener all together, the problem will be gone

I only hope I'm not missing something.

rkhachatryan replied:
> There is an inconsistency issue from perspective of InputChannels threads. But they are accessing AlternatingCheckpointBarrierHandler only via the notifyBarrierReceived and notifyBufferReceived

If I got your point right,
`InputChannels` threads notify `CheckpointBarrierUnaligner.ThreadSafeUnaligner`, not `AlternatingCheckpointBarrierHandler`.
And `ThreadSafeUnaligner` doesn't know about the `AlternatingCheckpointBarrierHandler`.

There is "inconsistency" in the sense that `AlternatingCheckpointBarrierHandler` can "lag behind" `ThreadSafeUnaligner` but it shouldn't cause any problems.

pnowojski replied:
> There is "inconsistency" in the sense that AlternatingCheckpointBarrierHandler can "lag behind" ThreadSafeUnaligner but it shouldn't cause any problems.

Yes, me and Arvid think and hope that's the case. (I've synced with him offline)

AHeise commented on flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/io/AlternatingCheckpointBarrierHandlerTest.java:
```diff
@@ -0,0 +1,212 @@
+	}
+
+	@Test
+	public void testAlternation() throws Exception {
```
I guess this test is super fast (<1s). If it takes longer, please reduce `numChannels` a bit.

rkhachatryan replied:
Checked locally, it's just a matter of milliseconds.

AHeise commented on flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/io/AlternatingCheckpointBarrierHandlerTest.java:
```diff
@@ -0,0 +1,212 @@
+
+	private static CheckpointedInputGate buildGate(TestInvokable target, int numChannels) {
+		SingleInputGate gate = new SingleInputGateBuilder().setNumberOfChannels(numChannels).build();
+		TestInputChannel[] channels = new TestInputChannel[numChannels];
+		for (int i = 0; i < numChannels; i++) {
+			channels[i] = new TestInputChannel(gate, i);
+		}
+		gate.setInputChannels(channels);
```
nit: could be a oneliner with (Int)Stream.

rkhachatryan replied:
In this case it didn't improve readability.

AHeise commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/AlternatingCheckpointBarrierHandler.java:
```diff
@@ -0,0 +1,76 @@
+import org.apache.flink.runtime.io.network.api.CheckpointBarrier;
+import org.apache.flink.runtime.jobgraph.tasks.AbstractInvokable;
+
+class AlternatingCheckpointBarrierHandler extends CheckpointBarrierHandler {
```
You most defintively need to delegate `getBufferReceivedListener`.

rkhachatryan replied:
Good point!

AHeise commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/SubtaskCheckpointCoordinatorImpl.java:
```diff
@@ -221,16 +221,10 @@ private void prepareInflightDataSnapshot(long checkpointId) throws IOException {
-			channelWrittenFuture = FutureUtils.completedVoidFuture();
-		}
+	private void finishAndReportAsync(Map<OperatorID, OperatorSnapshotFutures> snapshotFutures, CheckpointMetaData metadata, CheckpointMetrics metrics, CheckpointOptions options) {
+		Future<?> channelWrittenFuture = unalignedCheckpointEnabled && !options.getCheckpointType().isSavepoint() ?
```
Same on ternary.

rkhachatryan replied:
Reverted to minimize changes in this PR.

rkhachatryan commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/RemoteInputChannel.java:
```diff
@@ -560,8 +559,8 @@ public void onBuffer(Buffer buffer, int sequenceNumber, int backlog) throws IOEx
-			final CheckpointBarrier notifyReceivedBarrier;
-			final Buffer notifyReceivedBuffer;
+			final Optional<CheckpointBarrier> notifyReceivedBarrier;
+			final Optional<Buffer> notifyReceivedBuffer;
```
Yes, only for symmetry.

rkhachatryan commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/InputChannel.java:
```diff
@@ -291,7 +291,7 @@ public int unsynchronizedGetNumberOfQueuedBuffers() {
 	 */
 	@Nullable
-	protected CheckpointBarrier parseCheckpointBarrierOrNull(Buffer buffer) throws IOException {
+	CheckpointBarrier parseCheckpointBarrierOrNull(Buffer buffer) throws IOException {
```
This change is necessary to create `TestInputChannel` (in a different package).

AHeise replied:
But you are decreasing visibility here...

rkhachatryan replied:
You're right, reverting this change.

rkhachatryan commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/RemoteInputChannel.java:
```diff
@@ -551,7 +551,7 @@ void onSenderBacklog(int backlog) throws IOException {
 
 	public void onBuffer(Buffer buffer, int sequenceNumber, int backlog) throws IOException {
-		boolean recycleBuffer = true;
+		boolean enqueued = false;
```
> In case of some parse exception, the whole TM is restarted anyways.

That's true for now, but it can change in the future.

> renaming and inversion is unnecessary

This is to give more context to prevent future bugs.

AHeise replied:
While `enqueued` is more semantic, I'd assume that the technical `recycleBuffer` should be more explicit concerning bugs.

rkhachatryan replied:
Reverted to minimize changes in this PR.

rkhachatryan commented:
Thanks for the feedback @AHeise ,
I added channel tests and implemented `getBufferReceivedListener`.
Please take a look.

AHeise commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/InputChannel.java:
```diff
@@ -300,7 +300,7 @@ protected CheckpointBarrier parseCheckpointBarrierOrNull(Buffer buffer) throws I
 		// we can further improve to avoid double deserialization in the future.
 		buffer.setReaderIndex(0);
-		return event.getClass() == CheckpointBarrier.class ? (CheckpointBarrier) event : null;
+		return event instanceof CheckpointBarrier ? (CheckpointBarrier) event : null;
```
Do we even have any subclass here?

rkhachatryan replied:
No, but if there will be one then we'll need to update every such place.

rkhachatryan replied:
Reverted to minimize changes in this PR.

AHeise commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/SingleInputGateFactory.java:
```diff
@@ -124,7 +124,7 @@ public SingleInputGate create(
 			bufferDecompressor = new BufferDecompressor(networkBufferSize, compressionCodec);
 		}
 
-		SingleInputGate inputGate = new SingleInputGate(
```
Are the last three hotfix commits even used in the PR? (Introducing interfaces for gates/channels?)

pnowojski replied:
I second this question. On it's own, I think they are doing more harm then good (adding extra code that we would have to maintain).

rkhachatryan replied:
Yes, I added and used test implementations for the extracted interfaces.
IMO, extracting interface is a trade-off (and it often pays off). Particularly in this case, with many tests and unnecessary complex (for tests) class:
```
public SingleInputGateImpl(
    String owningTaskName,
    int gateIndex,
    IntermediateDataSetID consumedResultId,
    final ResultPartitionType consumedPartitionType,
    int consumedSubpartitionIndex,
    int numberOfInputChannels,
    PartitionProducerStateProvider partitionProducerStateProvider,
    SupplierWithException<BufferPool, IOException> bufferPoolFactory,
    @Nullable BufferDecompressor bufferDecompressor) {
```

rkhachatryan replied:
Reverted to minimize changes in this PR.

pnowojski commented on flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/consumer/TestInputGate.java:
```diff
@@ -0,0 +1,174 @@
+import java.util.Timer;
+import java.util.concurrent.CompletableFuture;
+
+class TestInputGate implements SingleInputGate {
```
Why do you need this class? Couldn't you use:
```
new SingleInputGateBuilder().setNumberOfChannels(numberOfChannels).build();
```
?

Besides `TestInputGate` being a bit dangerous - some methods are implemented incorrectly without throwing an exception, it will add quite a bit of extra maintenance whenever we will going to change `SingleInputGate` interface.

rkhachatryan replied:
Reverted to minimize changes in this PR.

pnowojski commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/InputChannel.java:
```diff
@@ -290,17 +288,16 @@ public int unsynchronizedGetNumberOfQueuedBuffers() {
 	 */
-	@Nullable
-	CheckpointBarrier parseCheckpointBarrierOrNull(Buffer buffer) throws IOException {
+	Optional<CheckpointBarrier> parseCheckpointBarrier(Buffer buffer) throws IOException {
```
This is irrelevant change to this commit. Changing `null` to `Optional` hides the real fix/change, so can you please split it into two commits?

Side note, `null` -> `Optional` change can have negative performance impact (unlikely, but possible and previous such changes were going through extra benchmark tests), so it would be good to benchmark it. However we don't have benchmarks with unaligned checkpoints at the moment. Because of that, I would leave it as it is (`@Nullable`) before the feature freeze.

rkhachatryan replied:
Reverted to minimize changes in this PR.

rkhachatryan commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/SubtaskCheckpointCoordinatorImpl.java:
```diff
@@ -152,7 +152,7 @@ public void checkpointState(
 
 		// Step (3): Prepare to spill the in-flight buffers for input and output
-		if (unalignedCheckpointEnabled) {
+		if (unalignedCheckpointEnabled && !options.getCheckpointType().isSavepoint()) {
```
Added `SubtaskCheckpointCoordinatorTest.testSkipChannelStateForSavepoints`.
Thanks for pointing it out!

rkhachatryan commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/RemoteInputChannel.java:
```diff
@@ -573,6 +573,7 @@ public void onBuffer(Buffer buffer, int sequenceNumber, int backlog) throws IOEx
 
 				wasEmpty = receivedBuffers.isEmpty();
 				receivedBuffers.add(buffer);
+				enqueued = true;
```
Reverted to minimize changes in this PR.

rkhachatryan commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/AlternatingCheckpointBarrierHandler.java:
```diff
@@ -0,0 +1,83 @@
+			activeHandler = alignedHandler;
+		}
+		activeHandler.processBarrier(receivedBarrier, channelIndex);
+	}
```
I've changed the logic of `AlternatingCheckpointBarrierHandler` to abort the previous checkpoint and reset the previous handler.

rkhachatryan commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/AlternatingCheckpointBarrierHandler.java:
```diff
@@ -0,0 +1,83 @@
+	}
+
+	@Override
+	public void processEndOfPartition() throws Exception {
+		alignedHandler.processEndOfPartition();
+		unalignedHandler.processEndOfPartition();
+	}
```
Added `AlternatingCheckpointBarrierHandlerTest.testEndOfPartition `,
thanks for pointing it out.

rkhachatryan commented:
Thanks for the feedback and sorry for the messy commits.
I removed refactorings and unrelated bug fixes, please take a look.


> Are the non hotfix commits from this PR independent? Could they be re-ordered to avoid squashing?

I didn't figure out the right order which would provide correctness so I squashed the remaining commits.

> Also aren't we missing an ITCase testing one way or another that aligned savepoints do not stare any data?

I think this was added in #12000

rkhachatryan commented:
@flinkbot run azure

pnowojski commented on flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/consumer/LocalInputChannelTest.java:
```diff
@@ -456,6 +463,23 @@ public void testUnblockReleasedChannel() throws Exception {
+	@Test
+	public void testNoNotifyOnSavepoint() throws IOException {
+		TestBufferReceivedListener listener = new TestBufferReceivedListener();
+		LocalInputChannel channel = new LocalInputChannel(
+			new SingleInputGateBuilder().build(),
+			0,
+			new ResultPartitionID(),
+			new ResultPartitionManager(),
+			new TaskEventDispatcher(),
+			new TestCounter(),
+			new TestCounter());
```
nit:
```
InputChannelBuilder
  .newBuilder()
  .buildLocalChannel(new SingleInputGateBuilder().build())
```
?

pnowojski commented on flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/consumer/LocalInputChannelTest.java:
```diff
@@ -456,6 +463,23 @@ public void testUnblockReleasedChannel() throws Exception {
+		CheckpointBarrier barrier = new CheckpointBarrier(123L, 123L, new CheckpointOptions(SAVEPOINT, CheckpointStorageLocationReference.getDefault()));
+		channel.notifyPriorityEvent(new BufferConsumer(toBuffer(barrier).getMemorySegment(), FreeingBufferRecycler.INSTANCE, getDataType(barrier)));
+		channel.checkError();
+		assertTrue(listener.notifiedOnBarriers.isEmpty());
```
frankly, it's not that unexpected that it's empty, since `listener` was never passed anywhere? 😈 

(you forgot to register it to `SingleInputGate`?)

pnowojski commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/InputProcessorUtil.java:
```diff
@@ -109,10 +109,17 @@ private static CheckpointBarrierHandler createCheckpointBarrierHandler(
-						numberOfInputChannelsPerGate.toArray(),
-						channelStateWriter,
-						taskName,
+					return new AlternatingCheckpointBarrierHandler(
+						new CheckpointBarrierAligner(
+							taskName,
+							channelIndexToInputGate,
+							inputGateToChannelIndexOffset,
+							toNotifyOnCheckpoint),
+						new CheckpointBarrierUnaligner(
+							numberOfInputChannelsPerGate.toArray(),
+							channelStateWriter,
+							taskName,
+							toNotifyOnCheckpoint),
```
There are two methods in `CheckpointedInputGate` which are implemented in not very object oriented design (I guess they should be re-implemented) that would stop working with this change (I hope some tests will fail because of that):
```
	public void spillInflightBuffers(
			long checkpointId,
			int channelIndex,
			ChannelStateWriter channelStateWriter) throws IOException {
		if (((CheckpointBarrierUnaligner) barrierHandler).hasInflightData(checkpointId, channelIndex)) {
			inputGate.getChannel(channelIndex).spillInflightBuffers(checkpointId, channelStateWriter);
		}
	}

	public CompletableFuture<Void> getAllBarriersReceivedFuture(long checkpointId) {
		return ((CheckpointBarrierUnaligner) barrierHandler).getAllBarriersReceivedFuture(checkpointId);
	}
```
I guess `hasInflightData()` and `getAllBarriersReceivedFuture()` should be pulled to `CheckpointBarrierHandler` interface (separate commit?)?

rkhachatryan replied:
Good point! 
Moving up the methods.

pnowojski commented:
One more thing. I think we are missing here an ITCase, that would verify user can modify job graph (changing record schema for example) after taking a savepoint in unaligned checkpoint mode.

rkhachatryan commented:
Thanks for the feedback @pnowojski, I've addressed the remaining issues.

Regarding the IT case, I think it's out of the scope of this PR because it only fixes savepoint creation in UC mode without changing the snapshot format or recovery logic.

[This thread](https://github.com/apache/flink/pull/11948#discussion_r424570447) remains open, could you please revisit it @AHeise and/or @pnowojski?

pnowojski commented:
@flinkbot run azure

rkhachatryan commented:
@flinkbot run azure

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/RemoteInputChannel.java:
```diff
@@ -469,7 +469,9 @@ public void onBuffer(Buffer buffer, int sequenceNumber, int backlog) throws IOEx
 			if (notifyReceivedBarrier != null) {
 				receivedCheckpointId = notifyReceivedBarrier.getId();
-				listener.notifyBarrierReceived(notifyReceivedBarrier, channelInfo);
+				if (notifyReceivedBarrier.isCheckpoint()) {
```
I guess the above `receivedCheckpointId = notifyReceivedBarrier.getId();` should also be placed inside this condition, because `receivedCheckpointId` only works for checkpoint now.

pnowojski replied:
But the condition works the other way around? If we bump/increase `receivedCheckpointId` here, it shouldn't matter. Spilling from input channels starts only if `lastRequestedCheckpointId` is bumped over the `receivedCheckpointId`.

zhijiangW replied:
Yes, it should be no problem here. I forgot the truth that the savepoint id is consecutive incremental based on checkpoint id.

rkhachatryan replied:
> above receivedCheckpointId < lastRequestedCheckpointId might be invalid 

Yes, the intention was to invalidate this condition. 
So the next data buffer after a *savepoint* barrier doesn't notify the listener and doesn't go to `ChannelStateWriter`.

>  to make some unnecessary notifyReceivedBuffer to be spilled in listener.

If the condition is invalidated then the listener is NOT notified.

zhijiangW commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/AlternatingCheckpointBarrierHandler.java:
```diff
@@ -0,0 +1,102 @@
+
+	@Override
+	public void processEndOfPartition() throws Exception {
+		alignedHandler.processEndOfPartition();
```
A bit confuse here why we need call both handlers, should only the active handler work at the same time?

rkhachatryan replied:
I this case both handlers need to update their number of open input channels.
